lib/puppet/environments.rb

```diff
@@ -447,5 +451,27 @@ def expired?
         Time.now > @ttl
       end
     end
+
+    class FileMtimeEntry < Entry
+      def initialize(env, path)
+        super value
+        @mtime   = Time.now
+        @evicted = false
+        @thread  = Thread.new do
```

Comment:
would still evict multiple times per catalog compilation on average. i guess it would help performance a lot, but i still prefer evicting when deployment happens. the only problem is that i'm hijacking 0 value and it would be undocumented behavior so whenever we stop touching environment.conf on deployment it would stop evicting - that needs to be more explicit.